dali/operators/random/rng_base_gpu.cuh

```diff
@@ -0,0 +1,125 @@
+  }
+  const int block_size_;
+  const int max_blocks_;
+  curand_states randomizer_;
+  kernels::memory::KernelUniquePtr<BlockDesc> block_descs_gpu_;
+  kernels::memory::KernelUniquePtr<BlockDesc> block_descs_cpu_;
+};
+
+namespace {
+
+template <typename Out, typename Dist, bool DefaultDist = false>
+__global__ void RNGKernel(BlockDesc* __restrict__ block_descs,
+                          curandState* __restrict__ states,
+                          const Dist* __restrict__ dists, int nblocks) {
+  int block_size = blockDim.x * blockDim.y;
+  int local_tid = blockDim.x * threadIdx.y + threadIdx.x;
+  int64_t global_tid = blockIdx.y * block_size + local_tid;
+  auto rng = states + global_tid;
+  int blk_stride = blockDim.y * gridDim.y;
+  int blk = blockIdx.y * blockDim.y + threadIdx.y;
+  for (; blk < nblocks; blk += blk_stride) {
+    auto desc = block_descs[blk];
+    auto start = static_cast<Out*>(desc.start);
+    auto block_end = start + desc.size;
+    Dist dist = DefaultDist ? Dist() : dists[desc.sample_idx];
+    for (auto out = start + threadIdx.x; out < block_end; out += blockDim.x) {
+      *out = ConvertSat<Out>(dist(rng));
+    }
+  }
+}
+
+}  // namespace
+
+template <typename Backend, typename Impl>
+template <typename T, typename Dist>
+void RNGBase<Backend, Impl>::RunImplTyped(workspace_t<GPUBackend> &ws) {
+  static_assert(std::is_same<Backend, GPUBackend>::value);
+  auto out_view = view<T>(ws.template OutputRef<GPUBackend>(0));
+  int nsamples = out_view.shape.size();
+  auto blocks_cpu = backend_data_.block_descs_cpu_.get();
+  auto blocks_gpu = backend_data_.block_descs_gpu_.get();
+  auto rngs = backend_data_.randomizer_.states();
+  int block_sz = backend_data_.block_size_;
+  int max_nblocks = backend_data_.max_blocks_;
+  int blockdesc_count = -1;
+  blockdesc_count = SetupBlockDescs(
+    blocks_cpu, block_sz, max_nblocks, out_view, ws.stream());
+
+  int64_t blockdesc_max_sz = -1;
+  for (int b = 0; b < blockdesc_count; b++) {
+    int64_t block_sz = blocks_cpu[b].size;
+    if (block_sz > blockdesc_max_sz)
+      blockdesc_max_sz = block_sz;
+  }
+
+  cudaMemcpyAsync(blocks_gpu, blocks_cpu,
+                  sizeof(BlockDesc) * blockdesc_count, cudaMemcpyHostToDevice, ws.stream());
+  Dist* dists = This().template SetupDists<Dist>(nsamples, ws.stream());
+
+  dim3 blockDim;
+  dim3 gridDim;
+  const int kLargeVolumeThreshold = 4 * block_sz;
+  if (blockdesc_max_sz < kLargeVolumeThreshold) {  // small volumes
+    // A CUDA block processes several logical blocks
+    blockDim.x = std::min<int>(block_sz, blockdesc_max_sz);
+    blockDim.y = std::min<int>(blockdesc_count, block_sz / blockDim.x);
+    gridDim.x = 1;
+    gridDim.y = div_ceil(blockdesc_count, blockDim.y);
+  } else {  // large volumes
+    // A CUDA block processes a single logical block
+    blockDim.x = std::min<int>(blockdesc_max_sz, block_sz);
+    blockDim.y = 1;
+    gridDim.x = 1;
+    gridDim.y = blockdesc_count;
+  }
```

Comment:
I think this would serve both cases:
```suggestion
  blockDim.x = std::min<int>(block_sz, blockdesc_max_sz);
  blockDim.y = std::min<int>(blockdesc_count, std::max<int>(1, block_sz / blockDim.x));
  gridDim.x = 1;
  gridDim.y = div_ceil(blockdesc_count, blockDim.y);
```